### DISCUSSION TITLE
next/image, how to set image width to 100%?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#18474] [Help] [answered]
I tried to define my image like this:

```js
<Image
  src={props.image.url}
  alt="Picture of the author"
  width={props.image.width}
  height={props.image.height}
  className="card__image"
  />

<style jsx>{`
  .card__image {
    width: 100%;
  }
`}</style>
```

But that won't work. The image is only displayed at full size. What should we do when we wan't a fluid width for the images, to use in a responsive grid etc?

**Comment by MathiasKandelborg** [comment]
I think your question overlaps with [the one I recently made](https://github.com/vercel/next.js/discussions/18457).

There's a link to where the styling is added in the next/image component.

**Comment by zeekrey** [comment, marked as answer]
In Version 10.0.1 they added the `layout` prop to the image component. Setting it to `responsive` sets the css max-/min-width to 100%. Maybe this works for your usecase.

**Reply by MathiasKandelborg** [reply]
The layout prop solves most of these issues.

**Reply by ghost** [reply]
Looks great, will test it once `10.0.1` is released. :)

**Reply by MathiasKandelborg** [reply]
@eivindml-solvr It released 3 days ago: https://github.com/vercel/next.js/releases/tag/v10.0.1

**Reply by urgn** [reply]
But layout is deprecated right now. What should I use to get it responsive?

**Reply by aranoe** [reply]
Yes, what is the alternative to `layout`?

**Reply by nannu-coder** [reply]
> Yes, what is the alternative to `layout`?
```
<Image
src={imgSrc}
alt="image"
onError={handleImgError}
sizes="100vw"
fill
 />
```

**Reply by goodnewsj62** [reply]
this works but remember to add position relative on  the parent,  for me that was what solved my issue...  i did it like so
```
<div style={{position:"relative"}}>
      <Image
      src={imgSrc}
      alt="image"
      onError={handleImgError}
      sizes="100vw"
      objectFit="cover"
      fill
       />
</div>
```

**Comment by NishargShah** [comment]
In 2023, every above solution is deprecated,

use below

```
<Image
  src="url"
  width={0}
  height={0}
  sizes="100vw"
  style={{ width: '100%', height: 'auto' }} // optional
/>
```

**Reply by 77pintu** [reply]
@NishargShah   Thanks so much for given solution

**Reply by hariria** [reply]
this makes no sense 😆

**Reply by NishargShah** [reply]
@hariria it's a workaround 😂

**Reply by pmespresso** [reply]
this is madness 😂 , but thank you

**Reply by negu63** [reply]
How did you find out, it's amazing LOL 🤣

**Reply by krush11** [reply]
![](https://github.githubassets.com/images/icons/emoji/unicode/1f923.png)
Modern problems require modern solutions 
Thanks!! Helped a lot!

I feel vercel should work on improving the usage `next/image` to become more flexible. Granted the current props are a necessary evil for the image optimization and decreasing content shift, but there needs to be a better fix on this

**Reply by c-desmarais** [reply]
This does feel like a workaround, but it is the only solution I see working. Much thanks for proposing it. 

I wish there was a better way to use the Image component to simply fill 100% of the width of its parent element.

**Reply by mehedirana** [reply]
Its working

**Reply by imyunisafitri** [reply]
this is the one

**Reply by peterdresslar** [reply]
Kindly inform all the AIs ;-)

**Reply by sumitpanwar10** [reply]
Bruh..never thought you would be here...anyways thanks sar

**Reply by shikhar1verma** [reply]
This doesn't make any sense. XD

But its working.

If its working lets not touch it. XD

**Reply by ivanivanov10101** [reply]
Yup, this works.

**Reply by KNHui** [reply]
07

**Reply by p4cino** [reply]
works perfectly for preview/blob

**Reply by swport** [reply]
wouldn't using width:100% is going to stretch a smaller image.?

**Reply by NishargShah** [reply]
`width: auto` is also an option

**Reply by Ingcamilo1020** [reply]
Thanks

**Reply by brianonchain** [reply]
Nice trick. For SVG, you can leave out "sizes" attribute. But, for PNG, you must have a sizes attribute or the image will become blurry.

**Reply by qingyao-tan** [reply]
great, thank you

**Comment by aranoe** [comment]
For me this works with tailwind:
``` jsx
   <div  className="relative h-20 w-full">
       <Image
         src={url}
         fill
       />
    </div>
```

Ofcourse you can use `style` or any other `className`. But the parent element has to have `position: relative;`.

NOTE: `next/image` with `fill` prop produces this style:
```jsx
<img 
    style="position:absolute;height:100%;width:100%;left:0;top:0;right:0;bottom:0;color:transparent"
     ...
 />
```

**Reply by donguyenba** [reply]
For me this works with tailwind:
```javascript
<div className="relative block h-[220px] w-full">
<Image src={image} alt="image" fill style={{ objectFit: 'cover' }} />
</div>
```

**Comment by ChetSocio** [comment]
Why the freaking hell is it necessary to use Image tag only . As a developer, we must have freedom to use img or figure tag based on requirements. It seems next js always limiting choice for developers for its egoistic approaches

**Reply by NafiurSiddiqui** [reply]
This is because all the dynamic sizing benefits you get out of the box. But I understand your point and I wish they will add more flexibility to the image component.

**Comment by surya-probo** [comment]
I can see that now the Image component has given the choice to pass only pure numbers for width and height. Even if it is optional, without passing height and width, it gives the error.

**Comment by NafiurSiddiqui** [comment]
Anyone wondering about it, you just set width to 100% on the className. I am using tailwind and just simply `w-full` which is basically css `width:100%` works like a charm.
Make sure you have a parent. e.g -
```javascript
<div className="relative rounded-md w-full borderTest">
        <NextImage src={image} alt={imageAlt} className="rounded-t-md w-full" />
        <CardBlogTop />
</div>
```

**Comment by donguyenba** [comment]
If use want use fill props you can work here:


```javascript
<div className="relative block h-[220px] w-full">
    <Image src={image} alt="image" fill style={{ objectFit: 'cover' }} />
</div>
```

**Reply by KakkoiDev** [reply]
100% TailwindCSS solution: `object-cover`
```
<div className="relative size-20">
    <Image src={image} alt="image" fill={true} className="object-cover" />
</div>
```
Documentation: https://tailwindcss.com/docs/object-fit

**Comment by echives** [comment]
now nextJS official document gives an answer in 《Responsive Image》 chapter, here is the link
[Responsive Images](https://nextjs.org/docs/pages/api-reference/components/image#responsive-images)
```
  <Image
    src={me}
    alt="Picture of the author"
    sizes="100vw"
    style={{
      width: '100%',
      height: 'auto',
    }}
  />
```
and you can wrap it using a parent `<div>` tag to set custom size

**Reply by NishargShah** [reply]
I gave it this solution a year ago https://github.com/vercel/next.js/discussions/18474#discussioncomment-5501724 :)

**Reply by echives** [reply]
@NishargShah  Thank you for saving us from struggling

**Reply by seancheung** [reply]
This works for imported image only, not an url
